src/i2p.cpp

```diff
@@ -0,0 +1,432 @@
+{
+    CSHA256 hasher;
+    hasher.Write(dest.data(), dest.size());
+    unsigned char hash[CSHA256::OUTPUT_SIZE];
+    hasher.Finalize(hash);
+
+    CNetAddr addr;
+    const std::string addr_str = EncodeBase32(hash, false) + ".b32.i2p";
+    if (!addr.SetSpecial(addr_str)) {
+        throw std::runtime_error(strprintf("Cannot parse I2P address: '%s'", addr_str));
+    }
+
+    return addr;
+}
+
+/**
+ * Derive the .b32.i2p address of an I2P destination (I2P-style Base64).
+ * @param[in] dest I2P destination.
+ * @return the address that corresponds to `dest`
+ * @throw std::runtime_error if conversion fails
+ */
+static CNetAddr DestB64ToAddr(const std::string& dest)
+{
+    const Binary& decoded = DecodeI2PBase64(dest);
+    return DestBinToAddr(decoded);
+}
+
+namespace sam {
+
+Sock::Sock() : m_socket(INVALID_SOCKET) {}
+
+Sock::Sock(SOCKET s) : m_socket(s) {}
+
+Sock::Sock(Sock&& other)
+{
+    Reset();
+    m_socket = other.m_socket;
+    other.m_socket = INVALID_SOCKET;
+}
+
+Sock::~Sock() { Reset(); }
+
+Sock& Sock::operator=(Sock&& other)
+{
+    Reset();
+    m_socket = other.m_socket;
+    other.m_socket = INVALID_SOCKET;
+    return *this;
+}
+
+SOCKET Sock::Get() const { return m_socket; }
+
+SOCKET Sock::Release()
+{
+    const SOCKET s = m_socket;
+    m_socket = INVALID_SOCKET;
+    return s;
+}
+
+void Sock::Reset() { CloseSocket(m_socket); }
+
+Session::Session(const fs::path& private_key_file,
+                 const CService& control_host,
+                 CThreadInterrupt* interrupt)
+    : m_private_key_file(private_key_file), m_control_host(control_host), m_interrupt(interrupt),
+      m_control_sock(INVALID_SOCKET)
+{
+}
+
+Session::~Session()
+{
+    Disconnect();
+}
+
+bool Session::Accept(Connection& accepted)
+{
+    static constexpr auto err_wait_begin = 1s;
+    static constexpr auto err_wait_cap = 1min;
+    auto err_wait = err_wait_begin;
+
+    while (!*m_interrupt) {
+        try {
+            Sock sock;
+
+            {
+                LOCK(m_mutex);
+                CreateIfNotCreatedAlready();
+                accepted.me = m_my_addr;
+                sock = StreamAccept();
+            }
+
+            while (!*m_interrupt) {
+                Sockets sockets;
+                sockets.recv.insert({sock.Get()});
+                sockets.err.insert({sock.Get()});
+
+                SocketEvents(sockets, MAX_WAIT_FOR_IO, *m_interrupt);
+
+                if (sockets.recv.count(sock.Get()) == 0 && sockets.err.count(sock.Get()) == 0) {
+                    // No incoming connections within MAX_WAIT_FOR_IO.
+                    continue;
+                }
+
+                const std::string& peer_dest =
+                    RecvUntilTerminator(sock.Get(), '\n', MAX_WAIT_FOR_IO, *m_interrupt);
+
+                accepted.peer = CService(DestB64ToAddr(peer_dest), Params().GetDefaultPort());
+                accepted.socket = sock.Release();
+
+                err_wait = err_wait_begin;
```

Comment:
I think this is unnecessary.

Reply:
You mean the incremental wait time or just this line?

Reply:
Just this line because the next line is a `return true;`

Reply:
You are right, removed!